shell/platform/darwin/ios/framework/Source/FlutterViewController.mm

```diff
@@ -1200,9 +1200,10 @@ - (void)startKeyBoardAnimation:(NSTimeInterval)duration {
       }
       completion:^(BOOL finished) {
         if (self.displayLink == currentDisplayLink) {
+          // Indicates the displaylink captured by this block is the original one,which also
+          // indicates the animation has not been interrupted from its beginning. Moreover,indicates
+          // the animation is over and there is no more animation about to exectute.
           [self invalidateDisplayLink];
-        }
-        if (finished) {
```

Comment:
Yep, this looks like a good change to me.  Were you able to reproduce the issue jmagman mentioned?  This will be tricky to write a test for.

Reply:
This will definitely be tricky to test.  I don't love the partial mocks we use for this now since we're not really testing the behavior, but that the code is being executed in the right order
https://github.com/flutter/engine/blob/2962099077b36704071802ea5595f4a016a1c214/shell/platform/darwin/ios/framework/Source/FlutterViewControllerTest.mm#L187-L190

This doesn't really test that the change is fixed, but you could add an expectation/verification that the engine `-updateViewportMetrics:` is called after `startKeyBoardAnimation`.
https://github.com/flutter/engine/blob/2962099077b36704071802ea5595f4a016a1c214/shell/platform/darwin/ios/framework/Source/FlutterViewControllerTest.mm#L178
https://github.com/flutter/engine/blob/2962099077b36704071802ea5595f4a016a1c214/shell/platform/darwin/ios/framework/Source/FlutterViewControllerTest.mm#L338

Reply:
This test passed on this diff and fails on `main`:
```diff
                   }];
+
+  XCTestExpectation* expectation = [self expectationWithDescription:@"update viewport"];
+  OCMStub([mockEngine updateViewportMetrics:flutter::ViewportMetrics()])
+      .ignoringNonObjectArgs()
+      .andDo(^(NSInvocation* invocation) {
+        [expectation fulfill];
+      });
   id viewControllerMock = OCMPartialMock(viewController);
   [viewControllerMock keyboardWillChangeFrame:notification];
   OCMVerify([viewControllerMock startKeyBoardAnimation:0.25]);
+  [self waitForExpectationsWithTimeout:5.0 handler:nil];
```

Reply:
I took the liberty of pushing this test since this bug fix is blocking an internal release. Hope you don't mind. 🙂 

https://github.com/flutter/engine/pull/29862/commits/ac4fa44a9d96f0acc2acd17f01e8cb85875852a6